### PR TITLE
feat: choose correct abi to decode evm transactions

### LINKED ISSUE #671: ️🛠 - Choose correct abi to decode evm transactions
### Task description

Currently we always try to decode tx data with the ISC ABI. To improve this, we should check if the recipient is either one of the known erc20 contracts or the isc contract and use the corresponding ABI then  

### Requirements

- [ ] Get correct ABI to decode transaction data

### PR DESCRIPTION
**Pull request by MarkNerdi**
## Summary

> Please summarize your changes, describing **what** they are and **why** they were made.

...

### Platforms

> Please select any platforms where your changes have been tested.

-   **Desktop**
    -   [ ] MacOS
    -   [ ] Linux
    -   [ ] Windows

### Instructions

> Please describe the specific instructions, configurations, and/or test cases necessary to **test and verify** that your changes work as intended.

...

## Checklist

> Please tick the following boxes that are relevant to your changes.

-   [ ] I have followed the contribution guidelines for this project
-   [ ] I have performed a self-review of my own code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have added or modified tests that prove my changes work as intended
-   [ ] I have verified that new and existing unit tests pass locally with my changes
-   [ ] I have verified that my latest changes pass CI workflows for testing and linting
-   [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: packages/shared/src/lib/core/wallet/actions/isAddressATrackedToken.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { NetworkId } from '@core/network/types'
+import { IProfile } from '@core/profile'
+
+export function isAddressATrackedToken(networkId: NetworkId, address: string, profile: IProfile): boolean {
```
Comment (Tuditi) [inline packages/shared/src/lib/core/wallet/actions/isAddressATrackedToken.ts]: Maybe this sounds better?
```suggestion
export function isTrackedTokenAddress(networkId: NetworkId, address: string, profile: IProfile): boolean {
```

File: packages/shared/src/lib/core/layer-2/utils/getTransferInfoFromTransactionData.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            if (nativeToken) {
+                return {
+                    tokenId: nativeToken.ID.data,
+                    rawAmount: nativeToken.amount,
+                }
+            } else {
+                return undefined
+            }
+        }
+        if (decoded?.name === 'transfer') {
```
Comment (Tuditi) [inline packages/shared/src/lib/core/layer-2/utils/getTransferInfoFromTransactionData.ts]:
```suggestion
        } else if (decoded?.name === 'transfer') {
```

File: packages/shared/src/lib/core/layer-2/utils/getTransferInfoFromTransactionData.ts
Comment (Tuditi) [inline packages/shared/src/lib/core/layer-2/utils/getTransferInfoFromTransactionData.ts:1]: Why do you need this function? You're not using it anywhere.

File: packages/shared/src/lib/core/activity/utils/generateActivityFromEvmTransaction.ts [hunk 3]
```diff
@@ -25,17 +35,26 @@ export async function generateActivityFromEvmTransaction(
...
-        const abiDecoder = new AbiDecoder(ISC_SANDBOX_ABI, provider)
-        const decoded = abiDecoder.decodeData(transaction.data as string)
-        if (decoded?.name === 'call') {
-            const inputs = decoded.inputs as IscCallMethodInputs
-
-            const nativeToken = inputs?.allowance?.nativeTokens?.[0]
-            if (nativeToken) {
-                tokenId = nativeToken.ID.data
-                rawAmount = nativeToken.amount
+        const isErc20 = isAddressATrackedToken(networkId, recipient.address, getActiveProfile())
```
Comment (Tuditi) [inline packages/shared/src/lib/core/activity/utils/generateActivityFromEvmTransaction.ts]: I would move the call to `getActiveProfile()` into the function. I don't see the added benefit of passing an extra parameter.

File: packages/shared/src/lib/core/activity/utils/generateActivityFromEvmTransaction.ts [hunk 4]
```diff
@@ -25,17 +35,26 @@ export async function generateActivityFromEvmTransaction(
...
+        const isIscContract = recipient.address === ISC_MAGIC_CONTRACT_ADDRESS
+
+        const abi = isErc20 ? ERC20_ABI : isIscContract ? ISC_SANDBOX_ABI : undefined
+
+        if (abi) {
+            const abiDecoder = new AbiDecoder(abi, provider)
+            const decoded = abiDecoder.decodeData(transaction.data as string)
+            if (decoded?.name === 'call') {
```
Comment (Tuditi) [inline packages/shared/src/lib/core/activity/utils/generateActivityFromEvmTransaction.ts]: Maybe a switch statement works better here?

**Review by Tuditi — CHANGES_REQUESTED** [review]
Left some comments. I also get an error whenever I try to send fUSDC. Additionaly, the activity is generated twice.

Error:
```
Error: Amount is negative
    at formatTokenAmountDefault (formatTokenAmountDefault.ts:12:1)
    at formatTokenAmountBestMatch (formatTokenAmountBestMatch.ts:27:1)
    at Object.update [as p] (TokenAmountTile.svelte:52:1)
    at update_slot_base (index.mjs:100:1)
    at Object.update [as p] (Text.svelte:176:1)
    at Object.update [as p] (Text.svelte:176:1)
    at update (index.mjs:1298:1)
    at flush (index.mjs:1258:1)
```

Balance change:
image

File: packages/shared/src/lib/core/layer-2/utils/getTransferInfoFromTransactionData.ts
Comment (MarkNerdi) [reply packages/shared/src/lib/core/layer-2/utils/getTransferInfoFromTransactionData.ts:1]: i extracted the function into this file. but forgot to use it lol

**Review by Tuditi — DISMISSED** [review]
LGTM! :fire:

**Review by Tuditi — APPROVED** [review]

**Review by Tuditi — APPROVED** [review]

**Pull request merged**